src/dataflow/src/sink/tail.rs

```diff
@@ -30,19 +35,29 @@ pub fn tail<G>(
     let mut tx = block_on(connector.tx.connect()).expect("tail transmitter failed");
     let mut packer = RowPacker::new();
     stream.sink(Pipeline, &format!("tail-{}", id), move |input| {
-        input.for_each(|_, rows| {
+        input.for_each(|_, batches| {
             let mut results: Vec<Row> = Vec::new();
-            for (row, time, diff) in rows.iter() {
-                let should_emit = if connector.strict {
-                    connector.frontier.less_than(time)
-                } else {
-                    connector.frontier.less_equal(time)
-                };
-                if should_emit {
-                    packer.push(Datum::Decimal(Significand::new(i128::from(*time))));
-                    packer.push(Datum::Int64(i64::cast_from(*diff)));
-                    packer.extend_by_row(row);
-                    results.push(packer.finish_and_reuse());
+            for batch in batches.iter() {
+                let mut cursor = batch.cursor();
+                while cursor.key_valid(&batch) {
+                    while cursor.val_valid(&batch) {
+                        let row = cursor.val(&batch);
+                        cursor.map_times(&batch, |time, diff| {
+                            let should_emit = if connector.strict {
+                                connector.frontier.less_than(time)
+                            } else {
+                                connector.frontier.less_equal(time)
+                            };
+                            if should_emit {
+                                packer.push(Datum::Decimal(Significand::new(i128::from(*time))));
+                                packer.push(Datum::Int64(i64::cast_from(*diff)));
+                                packer.extend_by_row(row);
+                                results.push(packer.finish_and_reuse());
+                            }
```

Comment:
I think there is a glitch here in that a batch can contain multiple timestamps, but is ordered by key rather than time. So the timestamps can bounce around among the valid timestamps for the batch.

Probably the right fix is to stash the `(time, row)` pair, sort that by time, and then ship the results.

Reply:
Done.